Support passing Carbon `Optional(T*)` to C++ `T*` parameter.

zygoloid opened the pull request:
We already did this translation in the other direction, but we had no
mapping from `Optional(T)` to anything, so round-tripping a nullable
pointer from C++ through Carbon and back to C++ was previously rejected.

chandlerc commented on toolchain/check/cpp/type_mapping.cpp:
```diff
@@ -164,6 +184,25 @@ static auto TryMapClassType(Context& context, SemIR::ClassType class_type)
+          return WrappedType{
+              .inner_type_id = context.types().GetTypeIdForTypeInstId(
+                  pointer_type->pointee_id),
+              .wrap_fn = [](Context& context, clang::QualType inner_type) {
+                return context.ast_context().getPointerType(inner_type);
+              }};
```
This is a `function_ref` to a lambda that outlives the lambda itself... In this case, maybe its ok because this is a stateless lambda? But I'm not sure?

If this is OK, will anything catch if this stops being OK? Could we rename this te `wrap_fn_ref` or something to make it less surprising when this doesn't work?

Or if this is not OK, maybe `std::function` and then `wrap_fn` seems good?

zygoloid replied:
Switched to a function pointer for now, to catch use of captures that won't live long enough, with a comment that we can change this to `std::function` if we ever find we need captures here.

chandlerc reviewed (commented):
Generally LG, but a question on one aspect of using `function_ref` inline...

zygoloid committed:
Support passing Carbon `Optional(T*)` to C++ `T*` parameter.

We already did this translation in the other direction, but we had no
mapping from `Optional(T)` to anything, so round-tripping a nullable
pointer from C++ through Carbon and back to C++ was previously rejected.

zygoloid committed:
Replace function_ref with function pointer to derisk capture lifetimes.

chandlerc reviewed (approved):
LG, thanks!
